receiver/bigipreceiver/scraper.go

```diff
@@ -0,0 +1,299 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bigipreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/bigipreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/bigipreceiver/internal/metadata"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/bigipreceiver/internal/models"
+)
+
+// custom errors
+var (
+	errClientNotInit    = errors.New("client not initialized")
+	errScrapedNoMetrics = errors.New("failed to scrape any metrics")
+)
+
+// bigipScraper handles scraping of Big-IP metrics
+type bigipScraper struct {
+	client   client
+	logger   *zap.Logger
+	cfg      *Config
+	settings component.TelemetrySettings
+	mb       *metadata.MetricsBuilder
+}
+
+// newScraper creates an initialized bigipScraper
+func newScraper(logger *zap.Logger, cfg *Config, settings component.TelemetrySettings) *bigipScraper {
+	return &bigipScraper{
+		logger:   logger,
+		cfg:      cfg,
+		settings: settings,
+		mb:       metadata.NewMetricsBuilder(cfg.Metrics),
+	}
+}
+
+// start initializes a new big-ip client for the scraper
+func (s *bigipScraper) start(_ context.Context, host component.Host) (err error) {
+	s.client, err = newClient(s.cfg, host, s.settings, s.logger)
+	return
+}
+
+// scrape collects and creates OTEL metrics from a Big-IP environment
+func (s *bigipScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	// validate we don't attempt to scrape without initializing the client
+	if s.client == nil {
+		return pmetric.NewMetrics(), errClientNotInit
+	}
+
+	collectedMetrics := false
+
+	// initialize auth token
+	err := s.client.GetNewToken(ctx)
+	if err != nil {
+		return pmetric.NewMetrics(), err
+	}
+
+	var scrapeErrors scrapererror.ScrapeErrors
+	// scrape metrics for virtual servers
+	virtualServers, err := s.client.GetVirtualServers(ctx)
+	if err != nil {
+		scrapeErrors.AddPartial(0, err)
```

Comment:
This should be at least 1. If we can articulate how many metrics are lost, then use that number.

Reply:
@djaglowski so ultimately this means there was a failure at the resource level. so i know how many metrics per resource....but i don't know how many resources we failed to collect. are you still saying i should just put 1 in this case instead of 0 if the number can't be calculated properly?